locations.cpp

```diff
@@ -345,37 +360,32 @@ struct heptagon : cdata_or_heptagon {
   cell *c7;
   /** \brief associated generator of alternate structure, for Camelot and horocycles */
   heptagon *alt;
-  /** \brief connection table */
-  connection_table<heptagon> c;
-  // DO NOT add any fields after connection_table! (see tailored_alloc)

```

Comment:
Here's the benefit of the CRTP approach. We no longer have a named `c` data member that must-be-the-last-in-the-class. Instead, we have a base class `trailing_connection_table`, which provides a _member function_ `c()` that returns basically `(connection_table<heptagon>*)(this + 1)`.